### PR TITLE
feat: add script to find added or deleted declarations between commits

### PR DESCRIPTION
**Pull request by j-loreaux**
---

This will only work for commits after `scripts/declsIn.lean` is included in the lakefile. So, unfortunately, you can't test this on existing PRs. However, you can test it with commits on the current branch (i.e., make a new commit which adds / removes files / declarations, in leaf files so that you don't have to rebuild Mathlib; then use this tool against the older commit).

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by j-loreaux** [comment]
@alexjbest mentioned to me he is working on a more sophisticated tool. I'm not sure whether we'll want both or not, so I've requested his review.

The primary purpose of this tool is for dealing with PRs that do renaming, as in #8406

File: lakefile.lean [hunk 1]
```diff
@@ -43,6 +43,11 @@ lean_exe checkYaml where
   root := `scripts.checkYaml
   supportInterpreter := true
 
+/-- `lake exe declsIn` lists all declarations provided in the given files. -/
```
Comment (semorrison) [inline lakefile.lean:46]: Could you add instructions here, that users will want to use `scripts/decl-name-changes.sh`?

File: scripts/decl-name-changes.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
+
```
Comment (semorrison) [inline scripts/decl-name-changes.sh:2]: A comment explaining usage, with and without a command line argument?

Comment (mo271) [inline scripts/decl-name-changes.sh:2]:
```suggestion

set -e
```

Perhaps make this script exit on error. But then again the files declsOld.txt declsNew.txt might not be cleaned up in that case...

Comment (j-loreaux) [reply scripts/decl-name-changes.sh:2]: I haven't made it exit on error, but I have added a warning to the usage that early termination could cause git to end up in a detached head state.

File: scripts/declsIn.lean [hunk 3]
```diff
@@ -6,7 +6,11 @@ import Mathlib.Lean.CoreM
 open Lean
 
 /-- Given an array of modules, this function returns the sorted list of all non-blacklisted
-declarations provided by those modules. -/
+declarations provided by those modules.
+
+The script `decl-name-changes.sh` wraps this and uses `git` to detect which files have changed.
+Then it computes the `diff` between the arrays returned by this function.  In general, users should
```
Comment (mo271) [inline scripts/declsIn.lean:12]:
```suggestion
Then it computes the `diff` between the arrays returned by this function. In general, users should
```

**Comment by grunweg** [comment]
I was renaming declarations the other day - and this script would have been really helpful. Any changes this (or its successor) can be merged? (For me, an imperfect script is always better than none!)

**Comment by j-loreaux** [comment]
@alexjbest or @semorrison do you want to pull this branch, test it and (maintainer) merge or delegate? Likely it will be superseded at some point by Alex's `leaff` tool, but it seems like it should be helpful for the moment.

File: scripts/declsIn.lean [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+/-- Given an array of modules, this function returns the sorted list of all non-blacklisted
+declarations provided by those modules.
+
+The script `decl-name-changes.sh` wraps this and uses `git` to detect which files have changed.
+Then it computes the `diff` between the arrays returned by this function.  In general, users should
+prefer that instead of using this script directly. -/
+def getDeclsInPackagesNoBlacklist (pkgs : Array Name) : CoreM (Array Name) := do
+  let arr_decls ← pkgs.mapM fun pkg => do
+    (← Std.Tactic.Lint.getDeclsInPackage pkg).filterM (Functor.map (! ·) ∘ Lean.Name.isBlackListed)
+  return arr_decls.flatten.heapSort (toString · < toString ·) |>.deduplicateSorted
```
Comment (alexjbest) [inline scripts/declsIn.lean:17]: More out of curiosity but why heapsort? and this seems to do a lot of conversions to string all the time why not use `Lean.Name.lt`

File: scripts/declsIn.lean [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+/-- Given an array of modules, this function returns the sorted list of all non-blacklisted
+declarations provided by those modules.
+
+The script `decl-name-changes.sh` wraps this and uses `git` to detect which files have changed.
+Then it computes the `diff` between the arrays returned by this function.  In general, users should
+prefer that instead of using this script directly. -/
+def getDeclsInPackagesNoBlacklist (pkgs : Array Name) : CoreM (Array Name) := do
+  let arr_decls ← pkgs.mapM fun pkg => do
+    (← Std.Tactic.Lint.getDeclsInPackage pkg).filterM (Functor.map (! ·) ∘ Lean.Name.isBlackListed)
```
Comment (alexjbest) [inline scripts/declsIn.lean:16]: does this work?
```suggestion
    (← Std.Tactic.Lint.getDeclsInPackage pkg).filterM (Functor.map (!Lean.Name.isBlackListed .))
```

**Comment by urkud** [comment]
What's the plan here? Am I right that `leaff` is already or soon will be strictly more powerful?

**Pull request closed without merging**